### ISSUE TITLE
[Question] pg_advisory_lock not supported by CockroachDB

### ISSUE THREAD

**Issue by jvanalstine** [jackc/tern#35]
Greetings, 

This is my favorite migration tool. However, I just started a new project with CockroachDB, and on migration I get a warning 
```
Error initializing migrator:
  ERROR: unknown function: pg_advisory_lock() (SQLSTATE 42883)
```
It looks like there are a [couple places](https://github.com/jackc/tern/blob/073150ed15213725dc32e87a0c37fe8c28d876d7/migrate/migrate.go#L273) where the function is used in migrate.go. Is it your opinion that this lock function is critical to the library? It seems like cockroach always uses `SERIALIZABLE isolation`.

**Comment by jackc** [comment]
The advisory lock is to protect against multiple migrations being run at the same time. tern would run fine without it, but concurrently running migrations would have unforeseen consequences.

#31 is somewhat related. The solution I proposed there was to allow user provided SQL for lock and unlock. With that in place you could use some other means of locking such as creating a table and checking its existence.

**Comment by jvanalstine** [comment]
> The solution I proposed there was to allow user provided SQL for lock and unlock. 

I am new to cockroach, but I believe they will not support pg_advisory_lock anytime soon. So this solution would likely work for roach. From what I can tell, there are no locks in cockroach. Also, I am not able to find any recommendations with respect to concurrently applying schema migrations. It looks like it runs raft and is CP, so I think they can just be applied concurrently without much worry. 

Tern was mentioned [here](https://github.com/cockroachdb/cockroach/issues/13546#issuecomment-731702643) btw

**Comment by jackc** [comment]
> Also, I am not able to find any recommendations with respect to concurrently applying schema migrations. It looks like it runs raft and is CP, so I think they can just be applied concurrently without much worry.

What I mean is that two migration processes are run at the same time. You never would want that to happen. e.g. migration 1 is running twice at the same time. Now it is extremely likely that one would fail because adding or dropping a table or column can only be done once. But if it was a data migration the user might not be so lucky.

Practically speaking though, it is a very unlikely case that is being protected against. An optional flag to disable locking might be reasonable.

**Comment by jvanalstine** [comment]
> What I mean is that two migration processes are run at the same time. You never would want that to happen. e.g. migration 1 is running twice at the same time. Now it is extremely likely that one would fail because adding or dropping a table or column can only be done once. But if it was a data migration the user might not be so lucky.

Makes sense. 

> An optional flag to disable locking might be reasonable.
I think a feature flag here makes a lot of sense. Do you want to take a wack at it or should I?

**Comment by jvanalstine** [comment]
It seems like there ought to be a flag called something like `--no-lock` that cases [this](https://github.com/jackc/tern/blob/master/main.go#L520) bit of code to be skipped and replaced with a regular transaction.

**Comment by jvanalstine** [comment]
I added the flag [here](https://github.com/jvanalstine/tern/tree/lock-flag). Although, there is still a problem with multiple transactions trying to run on the same connection. 

```
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
```

**Comment by jackc** [comment]
That error indicates that transactions are still being used somewhere and an error occurs inside it.

**Comment by jvanalstine** [comment]
Yes I figured that out thank you. Another thing that could be done is using the strategy that [golang-migrate](https://github.com/golang-migrate/migrate) uses. If the url has postgresql:// as the protocol, then an advisory lock is applied like in this repo. If cockroach:// is the protocol, then a lock table is created

```
  schema_name |    table_name     | type  |     owner      | estimated_row_count | locality
--------------+-------------------+-------+----------------+---------------------+-----------
  public      | schema_lock       | table | platform_admin |                   0 | NULL
  public      | schema_migrations | table | platform_admin |                   1 | NULL
```

**Comment by jackc** [comment]
Yeah, that might work. Though I'm a little reluctant to use lock tables. The nice thing about advisory locks is they automatically get cleaned up if the migration connection dies unexpectedly.  Whereas the strategy golang-migrate uses can require manual cleanup in an error case. See https://github.com/golang-migrate/migrate/blob/aef2747761d2a59d87e36fa942d9ece5339a2c41/database/cockroachdb/cockroachdb.go#L199.

**Comment by dghubble** [comment]
I'd love to move from `migrate` to `tern` but hit this too. Is the plan to wait for CockroachDB to make a stubbed out non-functional advisory lock implementation?

**Comment by jackc** [comment]
@dghubble

The only potential solutions that have come up on the tern side is an option to disable or customize the locking logic. I would lean to customizing the lock logic as that would also allow disabling it if desiring.

But I'm not currently working on this and I don't know anyone who is.

**Comment by tmcnicol** [comment]
Would you be open to a change along these lines? Any idea what kind of customisation would you want to allow?

Would allowing users to opt into the go migrate style be acceptable?

**Comment by jackc** [comment]
@tmcnicol Which approach? Several have been mentioned on this thread.

**Comment by tmcnicol** [comment]
Sorry I was referring to the go-migrate option, of a lock table that people could opt into. Was thinking that we could get around the unlocking when the connection unexpectedly dies by using a heartbeat mechanism.

**Comment by jackc** [comment]
@tmcnicol Yes, that seems reasonable.

**Comment by ryanluu12345** [comment]
@jackc , I'm wondering if there are any in progress changes that support the opt-in lock table support that @tmcnicol mentioned in his comment above?

**Comment by jackc** [comment]
@ryanluu12345 Not that I know of.

**Comment by tmcnicol** [comment]
Time necessitated that I moved to a different library for my application. I
don’t think it would be a hard change but didn’t have the capacity sorry.

**Comment by muhlemmer** [comment]
> Yeah, that might work. Though I'm a little reluctant to use lock tables. The nice thing about advisory locks is they automatically get cleaned up if the migration connection dies unexpectedly.

I'm wondering if doing a [`SELECT ... FOR UPDATE`](https://www.postgresql.org/docs/current/sql-select.html#SQL-FOR-UPDATE-SHARE) on the migrations table would be a possible solution. Those locks also get automatically cleaned up.

For example doing 

```sql
SELECT * FROM public.schema_version FOR UPDATE;
```

Will provide the same level of locking. Concurrent runs of tern will be locked until the previous lock is released, either by rollback or commit.

**Comment by muhlemmer** [comment]
Or [table level locking](https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-TABLES) on `public.schema_version`.

**Comment by kclonts** [comment]
I have an MR that should fix this

https://github.com/jackc/tern/pull/124
